### PR TITLE
Feature/api

### PR DESCRIPTION
**Pull request by lucblender**
Added routes:
- getAccessibleAttributes 
- getOwnedEndpoints
- getUserList
- modifyUserPassword
- addUserAuthority
- removeUserAuthority
Plenty of minor fixes!

### CODE REVIEW

File: src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt [hunk 1]
```diff
@@ -21,51 +20,60 @@ object  EndpointManagementUtil{
...
+        return if(endpointParameters == null)
+            null
+        else
+            EndpointFriendlyName(endpointParameters.friendlyName)
     }
 
     fun getNode(endpointEntityRepository: EndpointEntityRepository, nodeRequest: NodeRequest): Node? {
-        val splittedTopic = nodeRequest.nodeTopic.split("/")
-        return endpointEntityRepository.findByIdOrNull(splittedTopic[0])?.endpoint?.nodes?.get(splittedTopic[1])
+        val splitTopic = nodeRequest.nodeTopic.split("/")
```
Comment (cm0x4D) [inline src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:46]: Check if the split operation gives at least 2 parts before accessing [0] and [1]

File: src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt [hunk 2]
```diff
@@ -21,51 +20,60 @@ object  EndpointManagementUtil{
...
+        return endpointEntityRepository.findByIdOrNull(splitTopic[0])?.endpoint?.nodes?.get(splitTopic[1])
     }
 
     fun getWotNode(endpointEntityRepository: EndpointEntityRepository, nodeRequest: NodeRequest, host: String): WotNode? {
+        val splitTopic = nodeRequest.nodeTopic.split("/")
```
Comment (cm0x4D) [inline src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:54]: Check if the split operation gives at least 2 parts before accessing [0] and [1]

File: src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt [hunk 3]
```diff
@@ -21,51 +20,60 @@ object  EndpointManagementUtil{
...
-
-        return JsonWotSerializationFormat.wotNodeFromCloudioNode(endpointEntity.endpoint, endpointEntity.id, splitedTopic[1], host)
+        return JsonWotSerializationFormat.wotNodeFromCloudioNode(endpointEntity.endpoint, endpointEntity.endpointUuid, splitTopic[1], host)
     }
 
     fun getObject(endpointEntityRepository: EndpointEntityRepository, objectRequest: ObjectRequest): CloudioObject? {
-        val splittedTopic = Stack<String>()
-        splittedTopic.addAll(objectRequest.objectTopic.split("/").toList().reversed())
-        val endpointEntity = endpointEntityRepository.findByIdOrNull(splittedTopic.pop())
+        val splitTopic = Stack<String>()
```
Comment (cm0x4D) [inline src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:64]: Handle EmptyStackException

File: src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt [hunk 4]
```diff
@@ -21,51 +20,60 @@ object  EndpointManagementUtil{
...
         }
         return null
-
     }
 
     fun getAttribute(endpointEntityRepository: EndpointEntityRepository, attributeRequest: AttributeRequest): Attribute? {
-        val splittedTopic = Stack<String>()
-        splittedTopic.addAll(attributeRequest.attributeTopic.split("/").toList().reversed())
-        val endpointEntity = endpointEntityRepository.findByIdOrNull(splittedTopic.pop())
+        val splitTopic = Stack<String>()
```
Comment (cm0x4D) [inline src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:82]: Handle EmptyStackException

File: src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt [hunk 5]
```diff
@@ -75,16 +83,61 @@ object  EndpointManagementUtil{
...
         val attribute = getAttribute(endpointEntityRepository, AttributeRequest(attributeSetRequest.attributeTopic))
 
+        //only set attribute if setpoint or parameter
         if (attribute == null)
             return ApiActionAnswer(false, "Attribute doesn't exist")
         else if(attribute.constraint != AttributeConstraint.SetPoint && attribute.constraint != AttributeConstraint.Parameter)
             return ApiActionAnswer(false, "Attribute is nor a SetPoint, neither a Parameter")
         else {
+            //send message to amq.topic queue
             rabbitTemplate.convertSendAndReceive("amq.topic",
```
Comment (cm0x4D) [inline src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt]: If you do not expect an answer, use the convertAndSend() method.

**Review by cm0x4D — CHANGES_REQUESTED** [review]

**Review by cm0x4D — APPROVED** [review]

File: src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt (hunk 1 above)
Comment (lucblender) [reply src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:46]: Fixed in 0b262d6, added CloudioApiException and so removed the ApiActionAnswer data class

File: src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt (hunk 3 above)
Comment (lucblender) [reply src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:64]: Fixed in 0b262d6, same comment as above

File: src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt (hunk 2 above)
Comment (lucblender) [reply src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:54]: Fixed in 0b262d6, same comment as above

File: src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt (hunk 4 above)
Comment (lucblender) [reply src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt:82]: Fixed in 0b262d6, same comment as above

File: src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt (hunk 5 above)
Comment (lucblender) [reply src/main/kotlin/ch/hevs/cloudio/cloud/apiutils/EndpointManagementUtil.kt]: Fixed in 0b262d6

**Pull request merged**